### PR TITLE
Wffhcohort 822 - Support transitive joins in spark implementation

### PR DESCRIPTION
**Pull request by msargentibm**
This PR changes the structure of `ContextDefintion` so that related data is joined in to a context using a user-specified `joinClause`. It may be useful to take a look at the `docs` changes to get a feel for what is happening before diving into the code.

`ContextRetriever` is probably a good place to start looking through the code since most of the logic changes live there. Most other changes involve moving, deleting, or renaming code.

### CODE REVIEW

File: docs/user-guide/spark-user-guide.md [hunk 1]
```diff
@@ -290,9 +306,33 @@ entity "**vocabulary**" {
...
+        },
+        {
+          "name": "observation",
+          "joinClause": "inner join observation on person.person_id = observation.person_id"
+        }
+      ]
+    }
+  ]
+}
+```
```
Comment (rickjstevens) [inline docs/user-guide/spark-user-guide.md:335]: Not necessarily a concern, but wanted to confirm an observation I had. It looks like each of these joinClause statements are meant to be independent of each other and all based on/relative to primaryDataType? Versus incremental and based on prior joinClause statements? Assume the code and produces the final SQL statement to use for a context is able to filter out redundant logic across multiple joinClause expressions?

Comment (msargentibm) [reply docs/user-guide/spark-user-guide.md:335]: Your observation is correct. The joins work in the same manner that they did before these changes. Each join is independent.

**Review by shelleyMurphy — APPROVED** [review]
Looks much cleaner, I'm a fan!

**Review by rickjstevens — APPROVED** [review]
Looks good - I like the added flexibility this will enable for future use of our cohort spark implementation.

**Pull request merged**